packages/fuels-types/src/bech32.rs

```diff
@@ -0,0 +1,154 @@
+use crate::errors::Error;
+use bech32::Variant::Bech32m;
+use bech32::{FromBase32, ToBase32};
+use core::str;
+use fuel_tx::{Address, Bytes32, ContractId};
+use std::fmt::{Display, Formatter};
+
+pub const FUEL_BECH32_HRP: &str = "fuel";
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum ContentType {
+    Address,
+    ContractId,
+}
+
+/// An address in the bech32 format consisting of a
+/// human-readable part (HRP), '1' as a separator, the
+///  public key hash encoded with a u5 charset, and a checksum.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Bech32 {
```

Comment:
A more type strong approach will allow you to write code as follows:
```rust
        let bech32 = Bech32::Address("fuel1the_hash_part".parse()?);
        println!("{}", bech32);
        let address: Address = bech32.into();

        let bech32 = Bech32::ContractId("fuel1the_hash_part".parse()?);
        println!("{}", bech32);
        let contract_id: ContractId = bech32.into();
```

the implementation might go as follows:

Have `Bech32Data` be just a simple record-like struct:
```rust
pub struct Bech32Data {
    pub hrp: String,
    pub hash: Bytes32,
}
```
Like the real-world counterpart, it has a `hrp` and the hash behind it.

Make it so that it can be converted into `ContractId` and `Address`:
```rust
impl From<Bech32Data> for Address {
    fn from(data: Bech32Data) -> Self {
        Self::new(*data.hash)
    }
}
impl From<Bech32Data> for ContractId {
    fn from(data: Bech32Data) -> Self {
        Self::new(*data.hash)
    }
}
```
Implement `FromStr` to be able to parse a string into `Bech32Data`:
```rust
impl FromStr for Bech32Data {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (hrp, pubkey_hash_base32, _) = bech32::decode(s)?;

        let pubkey_hash: [u8; Address::LEN] = Vec::<u8>::from_base32(&pubkey_hash_base32)?
            .as_slice()
            .try_into()?;

        Ok(Self {
            hrp,
            hash: Bytes32::from(pubkey_hash),
        })
    }
}
```

Don't cache the data for the needs of `Display` -- compute it every time (unless benchmarking shows this to be too prohibitive):
```rust
impl Display for Bech32Data {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let data_base32 = self.hash.to_base32();
        let encoding = bech32::encode(&self.hrp, &data_base32, Bech32m).unwrap();

        write!(f, "{}", encoding)
    }
}
```

Don't use an additional enum for types, use types for types:
```rust
struct Bech32Address(Bech32Data);
```
Make it so that you can construct both an `Address` and `ContractId` from it:
```rust
impl From<Bech32Address> for Address {
    fn from(address: Bech32Address) -> Self {
        address.0.into()
    }
}
impl From<Bech32Address> for ContractId {
    fn from(address: Bech32Address) -> Self {
        address.0.into()
    }
}
```

it can also be parsed from a string:
```rust
impl FromStr for Bech32Address {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Bech32Address(s.parse()?))
    }
}
```
Also make it display-able:
```rust
impl Display for Bech32Address {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
```
Repeat the above for `Bech32ContractId`.

Then unite them in a `Bech32` enum so you can pass both around through one type:
```rust
enum Bech32 {
    Address(Bech32Address),
    ContractId(Bech32ContractId),
}
```

preferably add some qol methods for extracting the hash and hrp:
```rust
impl Bech32 {
    pub fn hash(&self) -> Bytes32 {
        match self {
            Bech32::Address(address) => address.0.hash,
            Bech32::ContractId(contract_id) => contract_id.0.hash,
        }
    }
    pub fn hrp(&self) -> &str {
        match self {
            Bech32::Address(address) => &address.0.hrp,
            Bech32::ContractId(contract_id) => &contract_id.0.hrp,
        }
    }
}
```

and finally, make it so that an `Address` and a `ContractId` can be constructed from it:
```rust

impl From<Bech32> for Address {
    fn from(bench32: Bech32) -> Self {
        Self::new(*bench32.hash())
    }
}

impl From<Bech32> for ContractId {
    fn from(bench32: Bech32) -> Self {
        Self::new(*bench32.hash())
    }
}
```

tldr the differences:
1. the bech32 is a simple data struct that resembles the real-world bech encoding (hrp + hash)
2. content type is embedded in the bech type itself
3. you use rust-idiomatic conversions instead of creating a proliferation of helper methods (construct_x_from_str, construct_x)

don't have particularly strong feelings towards this solution, just wanted to give it a try since it rubbed me the wrong way.

Reply:
Looks better, let me give it a try